### ISSUE TITLE
ML-Based Data Quality Validation Pipeline for Scrapers

### ISSUE THREAD

**Issue by holden** [razrfly/eventasaurus#2923]
## Overview

Build a unified ML-based validation pipeline to catch data quality issues **before** events are inserted into the database. This addresses several recurring problems:

- **Category classification** is subjective and rule-based approaches don't scale
- **Geographic mismatches** (e.g., Australian venues classified as UK)
- **Fake cities/venues** created from scraped UI elements or bad geocoding
- **City boundary confusion** - determining what belongs to one city vs another
- **Cross-source event matching** - same event from multiple sources with slight variations

### Current State
- **No ML models** - All validation is rule-based/heuristic
- **VenueNameValidator** - Good pattern using Jaro similarity (0.0-1.0 scores)
- **CategoryExtractor** - Pattern matching + YAML mappings, no confidence scoring
- **Error categorization** - 10 standard categories via pattern matching
- **MetricsTracker** - Already supports confidence scoring for collisions

### Gaps Identified
| Area | Current | Gap |
|------|---------|-----|
| Categories | Rule-based patterns | No confidence scores, subjective mapping |
| Location | GPS + k-d tree | No city-country consistency validation |
| Venues | Jaro similarity | No detection of scraped UI elements |
| Events | External ID dedup | No cross-source semantic matching |
| Titles | None | No quality/truncation detection |

### Directory Structure
```
lib/eventasaurus_discovery/ml_validation/
├── pipeline.ex                    # Main orchestrator
├── validators/
│   ├── validator_behaviour.ex     # Behaviour definition
│   ├── category_validator.ex      # BERT-based category validation
│   ├── geographic_validator.ex    # City-country consistency
│   ├── venue_validator.ex         # Venue name quality
│   ├── title_validator.ex         # Event title quality
│   └── temporal_validator.ex      # Date/time sanity
├── models/
│   ├── model_server.ex            # GenServer for model management
│   ├── category_classifier.ex     # BERT category model wrapper
│   └── semantic_similarity.ex     # BERT similarity model
├── confidence/
│   ├── score_aggregator.ex        # Combine validation scores
│   └── thresholds.ex              # Configurable thresholds
└── reporting/
    ├── validation_result.ex       # Result struct
    └── metrics_integration.ex     # MetricsTracker integration
```

### Pipeline Flow
```
Scraped Event Data
       ↓
[ML Validation Pipeline]
       ↓
   ┌───────────────────────────────────────┐
   │ 1. Title Validator (quality check)    │
   │ 2. Geographic Validator (geo sanity)  │
   │ 3. Venue Validator (name quality)     │
   │ 4. Category Validator (BERT model)    │
   │ 5. Temporal Validator (date sanity)   │
   └───────────────────────────────────────┘
       ↓
[Score Aggregator]
       ↓
   ┌───────────────────────────────────────┐
   │ Overall Confidence: 0.85              │
   │ - Title: 0.95 ✓                       │
   │ - Location: 0.90 ✓                    │
   │ - Venue: 0.70 ⚠️                      │
   │ - Category: 0.88 ✓                    │
   │ - Temporal: 1.0 ✓                     │
   └───────────────────────────────────────┘
       ↓
[Decision Gate]
       ↓
   confidence >= 0.8 → Insert to DB
   0.5 <= confidence < 0.8 → Insert + flag for review
   confidence < 0.5 → Reject + log error
```

### Validator Behaviour
```elixir
defmodule EventasaurusDiscovery.MLValidation.ValidatorBehaviour do
  @callback validate(event_data :: map()) :: 
    {:ok, score :: float(), details :: map()} | 
    {:error, reason :: term()}

  @callback name() :: String.t()
  @callback priority() :: integer()  # Execution order
  @callback weight() :: float()      # Importance in aggregation
  @callback fallback?() :: boolean() # Has rule-based fallback?
end
```

### Phase 1: Foundation (Quick Wins - No ML) ⚡
**Timeline: 1-2 weeks**

These require no ML models and provide immediate value:

#### 1.1 Validation Pipeline Framework
- [ ] Create `ValidatorBehaviour` module
- [ ] Implement `Pipeline` orchestrator
- [ ] Create `ScoreAggregator` for combining scores
- [ ] Add `ValidationResult` struct
- [ ] Integrate with `Processor.process_source_data/3`

#### 1.2 Geographic Validator (Addresses AU→UK problem)
- [ ] Check city is actually in claimed country
- [ ] Use existing k-d tree (156K cities) for validation
- [ ] Distance threshold: city must be within 500km of country centroid
- [ ] **Example catch:** Melbourne, Australia ≠ UK

```elixir
defmodule EventasaurusDiscovery.MLValidation.Validators.GeographicValidator do
  @impl true
  def validate(%{venue_data: venue_data}) do
    city = venue_data[:city]
    country = venue_data[:country]
    {lat, lng} = venue_data[:coordinates]
    
    # Use CityResolver's k-d tree to validate
    case CityResolver.validate_city_country(city, country, lat, lng) do
      {:ok, :valid} -> {:ok, 1.0, %{validated: true}}
      {:ok, :mismatch, actual_country} -> 
        {:ok, 0.2, %{expected: country, actual: actual_country}}
      {:error, :unknown_city} -> 
        {:ok, 0.5, %{warning: "City not in database"}}
    end
  end
end
```

#### 1.3 Temporal Validator
- [ ] Event not in past
- [ ] Event not > 2 years in future
- [ ] `end_at > start_at` (if both present)
- [ ] Duration reasonable (not 0, not > 30 days for single events)

#### 1.4 Title Quality Validator
- [ ] Length > 5 characters
- [ ] Not all caps (unless acronym)
- [ ] Not garbled/encoding issues
- [ ] Not truncated (doesn't end with "...")
- [ ] Not common UI elements ("Show More", "Loading...")

#### 1.5 Venue Quality Validator (Enhance existing)
- [ ] Extend `VenueNameValidator` with blacklist
- [ ] Add minimum length check (> 3 chars)
- [ ] Detect numeric-only names ("00000")
- [ ] Detect common UI elements

#### 1.6 Category Confidence Scoring
- [ ] Add confidence field to `CategoryMapper` results
- [ ] Score by match type: exact (1.0), pattern (0.8), fallback (0.5)
- [ ] Track source reliability in confidence

#### 2.1 Add Dependencies
```elixir
# mix.exs
{:nx, "~> 0.7"},
{:exla, "~> 0.7"},
{:bumblebee, "~> 0.5"},
{:tokenizers, "~> 0.4"}
```

#### 2.2 ModelServer GenServer
- [ ] Load models on application startup
- [ ] Warm model cache
- [ ] Handle inference requests
- [ ] Graceful degradation if model unavailable
- [ ] Memory management (lazy loading option)

#### 2.3 Sentence-BERT Integration
- [ ] Load `all-MiniLM-L6-v2` via Bumblebee
- [ ] Implement embedding generation
- [ ] Create similarity scoring function
- [ ] Use for venue/title validation enhancement

```elixir
defmodule EventasaurusDiscovery.MLValidation.Models.SemanticSimilarity do
  def compute_similarity(text1, text2) do
    emb1 = get_embedding(text1)
    emb2 = get_embedding(text2)
    cosine_similarity(emb1, emb2)
  end
end
```

#### 3.1 Training Data Preparation
- [ ] Export verified events with categories from database
- [ ] Clean and normalize text (title + description)
- [ ] Balance dataset across categories
- [ ] Split: 80% train, 10% val, 10% test
- [ ] Handle multi-label (events can have multiple categories)

#### 3.2 Model Selection & Training
- [ ] Base model: `xlm-roberta-base` (multilingual for PL/EN/FR content)
- [ ] Add multi-label classification head
- [ ] Fine-tune on event data using Axon
- [ ] Evaluate on test set (target: 85%+ F1)

#### 3.3 Deploy Category Classifier
- [ ] Integrate with `CategoryValidator`
- [ ] A/B test against rule-based approach
- [ ] Monitor confidence distribution
- [ ] Gradual rollout (10% → 50% → 100%)

#### 4.1 Cross-Source Event Matching
- [ ] Use semantic similarity for title matching
- [ ] Combine with existing GPS + date matching
- [ ] Improve confidence scoring for collisions
- [ ] Reduce duplicate events across sources

#### 4.2 Location Context Extraction (NER)
- [ ] Extract city/country mentions from event text
- [ ] Validate against claimed venue location
- [ ] Catch "Melbourne, Australia" in text but UK venue

#### 4.3 Continuous Learning Loop
- [ ] Track validation outcomes
- [ ] Collect human corrections as training data
- [ ] Periodic model retraining (monthly)
- [ ] Monitor model drift

## Quick Wins Priority List 🎯

| Priority | Validator | Effort | Impact | Dependencies |
|----------|-----------|--------|--------|--------------|
| 1 | Geographic Validator | Low | High | Existing k-d tree |
| 2 | Temporal Validator | Low | Medium | None |
| 3 | Title Quality Validator | Low | Medium | None |
| 4 | Category Confidence | Low | High | Existing CategoryMapper |
| 5 | Venue Quality Enhancement | Low | Medium | Existing VenueNameValidator |
| 6 | Sentence-BERT Similarity | Medium | High | Bumblebee deps |
| 7 | Category Classifier | High | Very High | Training data + ML deps |

### 1. Category Classifier (Priority 1)
- **Model:** `xlm-roberta-base` fine-tuned
- **Task:** Multi-label classification
- **Input:** Event title + description
- **Output:** Category probabilities (0.0-1.0 per category)
- **Libraries:** Bumblebee + Axon

### 2. Semantic Similarity (Priority 2)
- **Model:** `sentence-transformers/all-MiniLM-L6-v2`
- **Task:** Cosine similarity between texts
- **Use cases:** 
  - Cross-source event matching
  - Venue deduplication
  - Title quality validation
- **Libraries:** Bumblebee for embeddings

### 3. Named Entity Recognition (Priority 3)
- **Model:** `xlm-roberta-base` for NER
- **Task:** Extract location entities
- **Use cases:**
  - Validate venue location claims
  - Detect location mismatches in text

## Success Metrics

| Metric | Current | Target |
|--------|---------|--------|
| Obviously incorrect data entering DB | Unknown | 90% reduction |
| Geographic mismatches (AU→UK type) | Unknown | 99% caught |
| Fake city/venue creation | ~5-10/week | 80% reduction |
| Category assignment confidence | No scoring | 100% have scores |
| Cross-source duplicate detection | ~70% | 90%+ |

## Risks & Mitigations

| Risk | Impact | Mitigation |
|------|--------|------------|
| Model inference latency | Slower scrapers | Async/batch inference |
| Training data quality | Poor model | Use only verified events |
| Memory usage | OOM crashes | Lazy loading, quantization |
| Model unavailability | No validation | Rule-based fallback |

### 1. Processor Integration
```elixir
# lib/eventasaurus_discovery/sources/processor.ex
def process_source_data(events, source, scraper_name) do
  events
  |> Enum.map(&MLValidation.Pipeline.validate/1)
  |> Enum.filter(&above_confidence_threshold?/1)
  |> Enum.each(&process_event/1)
end
```

### 2. MetricsTracker Integration
```elixir
# Track validation results in job execution summaries
MetricsTracker.record_validation(job, %{
  overall_confidence: 0.85,
  validators: %{
    geographic: 0.90,
    temporal: 1.0,
    category: 0.88,
    venue: 0.70
  },
  decision: :accepted
})
```

### 3. Admin Dashboard
- [ ] New `/admin/validation-quality` dashboard
- [ ] Show confidence distributions
- [ ] List flagged events for review
- [ ] Track validation rejection rates

## Related Issues

- #2917 - Category classifier for events (this is the parent issue)
- Addresses: Geographic mismatches, fake cities, category subjectivity

## Labels

- `enhancement`
- `machine-learning`
- `data-quality`
- `scrapers`
- `priority-high`
